python/pyspark/sql/streaming.py

```diff
@@ -843,6 +844,169 @@ def trigger(self, processingTime=None, once=None, continuous=None):
         self._jwrite = self._jwrite.trigger(jTrigger)
         return self
 
+    def foreach(self, f):
+        """
+        Sets the output of the streaming query to be processed using the provided writer ``f``.
+        This is often used to write the output of a streaming query to arbitrary storage systems.
+        The processing logic can be specified in two ways.
+
+        #. A **function** that takes a row as input.
```

Comment:
Seems this variant is specific to Python. I thought we should better match how we support with Scala side.

Reply:
This is superset of what we support in scala. Python users are more likely to use simple lambdas instead of defining classes. But they may also want to write transactional stuff in python with open and close methods. Hence providing both alternatives seems to be a good idea.

Reply:
(including the response to https://github.com/apache/spark/pull/21477#discussion_r193631209) I kind of agree that it's a-okay idea but I think we usually provide a consistent API support so far unless it's language specific, for example, ContextManager, decorator in Python and etc.

Just for clarification, does Scala side support function only support too?

Also, I know attribute-checking way is kind of more like "Pythonic" way but I am seeing the documentation is already diverted between Scala vs Python. It costs maintaining overhead on the other hand.

Reply:
I mean, we could maybe consider the other ways but wouldn't it better to have the consistent support as the primary, and then see if the other ways are really requested by users? I think we could still incrementally add attribute-checking way or the lambda (or function to be more correct) way later.

Reply:
Python APIs anyways have slightly divergences from Scala/Java APIs in order to provide better experiences for Python users. For example, `StreamingQuery.lastProgress` returns an object of type `StreamingQueryProgress` in Java/Scala but returns a dict in python. Because python users are more used to dealing with dicts, and java/scala users (typed language) are more comfortable with structures). Similarly, in DataFrame.select, you can refer to columns in scala using `$"columnName"` but in python, you can refer to it as `df.columnName`. If we strictly adhere to pure consistency, then we cannot make it convenient for users in different languages. And ultimately convenience is what matters for the user experience. So its okay to have a superset of supported types in python compared to java/scala. 

Personally, I think we should also add the lambda variant to Scala as well. But that decision for Scala is independent of this PR as there is enough justification for add the lambda variant for

Reply:
I believe `$"columnName"` is more like a language specific feature in Scala and I think `df.columnName` is language specific to Python.

>  And ultimately convenience is what matters for the user experience.

Thing is, it sounded to me like we are kind of prejudging it..

>  I think we should also add the lambda variant to Scala as well.

+1

I am okay but I hope this shouldn't be usually done next time ..